Create ROADMAP file (#3746)
Azure/azure-service-operator: ROADMAP.md

# Roadmap

We track upcoming work on Azure Service Operator (ASO) using [GitHub Issues](https://github.com/Azure/azure-service-operator/issues?q=is%3Aissue+is%3Aopen).

## Release Cadence

We target a new release of ASO approximately every two months, though this may vary depending on the amount of work we have planned as well as external factors.

Our current release plan:

| Version | Estimated Release |
| ------- | ----------------- |
| [2.6.0](https://github.com/Azure/azure-service-operator/issues?q=is%3Aopen+is%3Aissue+milestone%3Av2.6.0) | Mid February 2024 |
| [2.7.0](https://github.com/Azure/azure-service-operator/issues?q=is%3Aopen+is%3Aissue+milestone%3Av2.7.0) | Mid April 2024 |
| 2.8.0 | Mid June 2024 |
| 2.9.0 | Mid August 2024 |

> Where linked, versions go to the list of open issues for that release.

Reviewer: ```suggestion
You can find a list of features and bugs that are planned to be fixed per release, in the respective links above.
```

> Where linked, versions go to the list of open issues for that release.

Reviewer: This feels like a start of a "planning" section.

> Any items note completed in time for one release will be carried over to the next, and may result in us pushing other items to a later release.

Reviewer: ```suggestion
Any work items that are not completed in time for a release will be carried over to the next one, and may result in delaying other work items to a later release as well.
```

> Partway through a release cycle, we'll review the list of issues assigned to the upcoming release and redistribute issues as needed.

Author: I'll clarify this. 

I also agree that we should perhaps use more than two upcoming milestones when planning, and limit our release plans to more reasonable sizes.

> If you're waiting on a particular resource or feature to be released, please comment on the relevant issue (or create a new issue) to let us know. We'll do our best to keep you updated on progress.

Reviewer 2: ```suggestion
If you're waiting on a particular resource or feature to be released, please comment on the relevant issue (or create a new issue if there isn't already one tracking the request) to let us know. We'll do our best to keep you updated on progress.
```

Author: Updated.

## Experimental Releases

> We publish an [experimental release](https://github.com/Azure/azure-service-operator/releases/tag/experimental) on a regular basis, suitable for testing and other non-production use cases. Such releases may not be stable and should not be used in production.

Reviewer 3: ```suggestion
We publish an [experimental release](https://github.com/Azure/azure-service-operator/releases/tag/experimental) for every push on main branch, suitable for testing and other non-production use cases. Such releases may not be stable and should not be used in production.
```

> We publish an [experimental release](https://github.com/Azure/azure-service-operator/releases/tag/experimental) on a regular basis, suitable for testing and other non-production use cases. Such releases may not be stable and should not be used in production.

Reviewer 3: ```suggestion
We publish an [experimental release](https://github.com/Azure/azure-service-operator/releases/tag/experimental) For every push on `main` branch, suitable for testing and other non-production use cases. Such releases may not be stable and should not be used in production.
```

Author: I didn't want to state `every push` in this document as people would take that as a binding promise, and we're already floating the idea of moving to _daily_ if we can avoid republishing when there have been no changes.

Reviewer: This is actually already merged and pushed for every push to main though

Reviewer: ```suggestion
We publish an [experimental release](https://github.com/Azure/azure-service-operator/releases/tag/experimental) on a regular basis, suitable for testing and other non-production use cases. Experimental releases are not be stable and should _not_ be used in production.
```

> ## Prior Releases

Reviewer: This is pretty far down, can we move it before the  "planning" section?

Author: This document isn't the canonical list of prior releases, it's a forward looking roadmap. 

I very nearly didn't include this section at all because it's backward looking, but figured it would be useful as a reference. I don't think it belongs earlier in the doc.

Prior GA releases of ASO v2:

| Version | Release Date |
| ------- | ------------ |
| [2.5.0](https://github.com/Azure/azure-service-operator/releases/tag/v2.5.0) | 7 Dec 2023 (1) |
| [2.4.0](https://github.com/Azure/azure-service-operator/releases/tag/v2.4.0) | 14 November 2023 |
| [2.3.0](https://github.com/Azure/azure-service-operator/releases/tag/v2.3.0) | 5 September 2023 |
| [2.2.0](https://github.com/Azure/azure-service-operator/releases/tag/v2.2.0) | 21 July 2023 |
| [2.1.0](https://github.com/Azure/azure-service-operator/releases/tag/v2.1.0) | 2 June 2023 |
| [2.0.0](https://github.com/Azure/azure-service-operator/releases/tag/v2.0.0) | 15 April 2023 |

(1) v2.5.0 had an abbreviated release cycle to get key items out before the 2023/2024 holiday season

## Issue Triage

We triage new issues weekly and assign most of them to an upcoming release milestone, based on our understanding of priority, complexity, and available resourcing.

## What about ASO v1?

> Azure Service Operator v1 is no longer under active development and we do not recommend it for new users. Bug and security fixes are still made to support existing users who are yet to migrate to v2.

Author: Added.
